docs/proposals/kubelet-eviction.md

```diff
@@ -0,0 +1,223 @@
+
+This document presents a specification for how the `kubelet` evicts pods when compute resources are too low.
+
+## Goals
+
+The node needs a mechanism to preserve stability when available compute resources are low.
+
+This is especially important when dealing with incompressible compute resources such
+as memory or disk.  If either resource is exhausted, the node would become unstable.
+
+The `kubelet` has some support for influencing system behavior in response to a system OOM by
+having the system OOM killer see higher OOM score adjust scores for containers that have consumed
+the largest amount of memory relative to their request.  System OOM events are very compute
+intensive, and can starve the node until the OOM killing process has completed.  In addition,
+the system is prone to return an unstable state since the containers that are killed due to OOM
+are just restarted, and the node may find itself soon starved of memory again. 
+
+Instead, we would prefer a system where the `kubelet` can pro-actively monitor for
+and prevent against total starvation of a compute resource, and in cases of where it
+could appear to occur, pro-actively fail a pod and terminate its containers so its workload
+can get moved and scheduled elsewhere when/if its backing controller creates a new pod.
+
+## Scope of proposal
+
+This proposal defines a pod eviction policy for reclaiming compute resources.
+
+In the first iteration, it focuses on memory; later iterations are expected to cover
+other resources like disk.  The proposal focuses on a simple default eviction strategy
+intended to cover the broadest class of user workloads.
+
+## Eviction Thresholds
+
+The `kubelet` will support the ability to specify eviction thresholds.  Each threshold 
+defines when a signal has met an administrator defined value that would impact node stability.
+If the threshold criteria has been met, the `kubelet` would take pro-active action to
+attempt to reclaim the starved compute resource associated with the signal.
+
+The `kubelet` will support two types of eviction thresholds.  A soft eviction threshold
+has an associated grace period, such that no action to reclaim the associated resource will
+be taken by the `kubelet` unless the grace period has been exceeded.  A hard eviction
+threshold has no grace period, and if observed, the `kubelet` will take immediate action
+to reclaim the associated starved resource.
+
+Initial support for the feature will focus on available memory, but other potential signals
+that may be added in the future include cpu load or available disk.
+
+If no eviction thresholds are specified, the `kubelet` will not do any eviction.
+
+To configure eviction thresholds, the following changes are recommended:
+
+```
+// EvictionSignal defines a signal that can trigger eviction of pods on a node.
+type EvictionSignal string
+
+const (
+	// EvictionMemoryAvailable, Memory available, in bytes
+	EvictionMemoryAvailable EvictionSignal = "memory.available"
```

Comment:
@dchen1107 @vishh - input appreciated on experience from borg/omega... do you find it more natural to define a signal based on memory.usage or what I thought of as memory.available here as basically free+buffers+cache?

I had looked to use memory.available since I thought it was more natural to configure across any machine size rather than needing to set a machine relative size which is what usage would require.